### PR TITLE
Feature: Gep-030 FanChart2Way

### PR DESCRIPTION
**Pull request by sam-m888**
...New FanChart consisting of both ascendants and descendants.

It can be checked-out there :
https://sourceforge.net/u/bubblegum00/gramps/ci/geps/gep-030-FanChart2Way/~/tree/

<snip>

I find it quite handy, so please let me know if you have any comments or suggestions.

NB: I heavily cleaned up the FanChart code on my way.

Regards,

Bastien Jacquet

https://sourceforge.net/p/gramps/mailman/message/32908110/
........................................................
Refactor fanchart for further modification

Fix radial text pos_start radian alignment

Simplify father's and mother's details getter

Rename draw_gradient to draw_gradient_legend

refactor prepare_background_box

Add option not to flip names

Add a display_format(self, person, num) function as proposed by Benny Malengier on 2012-12-13
It just returns name_display.name_formats[num][_F_FN](person.get_primary_name%28%29)

Add option to show the name on two lines

Allow for variable radius depending on generation

Switch to WORD_CHAR wrapping of name (ie word, and char if 0-length word)

Move rescaling tentative inside wrap_truncate_layout

Fix person_under_cursor bugs

Refactor root angle computation

Refactor code positionning the fan

Refactor personpos_at_angle

move implementation of person_under_cursor outside of FanChartBaseWidget class

Change draw_person to take angles in radians

Use same structure for innerring as for outerring

Uses cursor_to_polar and cursor_on_tranlation_dot

Slightly change person_under_cursor logic to return an "address" in the fan

Uses radian_in_bounds to compare angles modulo 2 PI

Fixup test on cursor over inner ring

Fix Center size for FanchartDesc

Fixup fanchart check up to last generation

Give same signature to draw_person

Refactor the common code of self.draw_person in a single function in Base class

Fix center box comment

Refactoring inside celladdress

Remove manual central box drawing since done with draw_person

Fixup draw_person color for duplicates

Use draw_person for central person too

Make __compute_angle and __rec_fill_data public for use in FanChart2Way

Add 2Way View

Rewrite create_map_rect_to_sector to allow bottom-outside-oriented text-arc

Allow to automatically right upside-up bottom arc-text

Correct icons for Fanchart2Way

Small code refactoring

Refactor code of fanchartdesc to use self.rootangle_rad

Rename change_slice to toggle_cell_state

Fanchart2Way code formating and changes

Small refactoring of fanchartdesc innerring fill data

Remove the name from the local temporary data structure

Remove the name from the local temporary data structure (in Fanchart2Way)

Change background gradient to follow the user-selected gradient colors

rename parentsroot to innerring

Some renaming for clearer code

Show last generation of partners in descendant fanchart

Show last partner in Fanchart2Way

Fanchart2Way : Add option to disable gradient on the background

Fixup flipupsidedownname parameter for gramplet usage of fancharts

Fixup twolinename parameter for gramplet usage of fancharts

Add FanChart2Way in available gramplets

Tentative fix for last view on Fanchart2Way

Show step-sibling in Fancharts context-menu

Fix overestimation of descendant halfdist

(SM) Trailing White spaces removed

(SM) Fix config box Table Grid

(SM) Move Icons gramps-fanchart2way to new location

(SM) Add Copyright for Bastien Jacquet

(SM) Fix BSDDB AttributeError NoneType object has no attr

(SM) Update patch to account for [bug 9771; fix missing right-click menu items](https://github.com/gramps-project/gramps/pull/268)

(Nick Hall/eno93) Fix set_text method takes the length of the utf-8, not the length of the unicode as the second parameter ((Gramps.py:3697): Pango-WARNING **: Invalid UTF-8 string passed to pango_layout_set_text())

### CODE REVIEW

**Comment by sam-m888** [comment]
Rebased on Gramps master

Testing needed as it modifies the existing fanchart view's improves text labels.

Few issues remain to be fixed:

 * Missing icon for view
 * Error when using in BSDDB mode (ok with DB-API)
 * Gtk-CRITICAL  gtk_widget_set_size_request: assertion 'width >= -1' failed
 * Pango-WARNING **: Invalid UTF-8 string passed to pango_layout_set_text()

fanchart2way-gramps-master-missing-icon-db-api-only

**Comment by sam-m888** [comment]
And it looks like according to Travis-CI I missed a few tailing whitespaces!
 * "ERROR - Trailing whitespace found in source file(s)"

**Comment by dsblank** [comment]
Wow! Great work from both of you! There were always some tiny issues with the original version that I could never resolve. This cleans up some of my original mistakes.

**Comment by eno93** [comment]
Very nice I like that I can both my ancestors and descendants on the one view.

**Comment by sam-m888** [comment]
Few issues remain to be fixed:

  - [ ] Gtk-CRITICAL  gtk_widget_set_size_request: assertion 'width >= -1' failed
  - [ ] Pango-WARNING **: Invalid UTF-8 string passed to pango_layout_set_text()

**Comment by daleathan** [comment]
Will this be part of Gramps 5.0?

**Comment by dsblank** [comment]
I hope it will make it into Gramps 5.0... I think it is good enough as is... just a couple of Gtk warnings left, right Sam?

**Comment by sam-m888** [comment]
@dsblank _just a couple of Gtk warnings left, right Sam?_

Having difficulty tracing down those GTK errors (I'm not good on GUI stuff), testing this morning showed up a few more issues below

  - [ ] With an empty family tree select(possibly drag) the empty chart view and get the following error:

```
22456: ERROR: grampsapp.py: line 140: Unhandled exception
Traceback (most recent call last):
  File "/home/grampsdev/aaSandbox/gramps/gramps/gui/widgets/fanchart.py", line 945, in on_mouse_up
    self.do_mouse_click()
  File "/home/grampsdev/aaSandbox/gramps/gramps/gui/widgets/fanchart2way.py", line 541, in do_mouse_click
    self.toggle_cell_state(self._mouse_click_cell_address)
  File "/home/grampsdev/aaSandbox/gramps/gramps/gui/widgets/fanchart2way.py", line 608, in toggle_cell_state
    self._compute_angles(*self.rootangle_rad_desc)
  File "/home/grampsdev/aaSandbox/gramps/gramps/gui/widgets/fanchartdesc.py", line 263, in _compute_angles
    data[2] = start
TypeError: 'tuple' object does not support item assignment

```
  - [ ] Do the same for the gramplet version (on an empty family tree)

```
184716: ERROR: grampsapp.py: line 140: Unhandled exception
Traceback (most recent call last):
  File "/home/grampsdev/aaSandbox/gramps/gramps/gui/widgets/fanchart.py", line 975, in on_drag_data_get
    id(self), person.get_handle(), 0)
AttributeError: 'NoneType' object has no attribute 'get_handle'

```

Also I've only tested on Linux (Ubuntu 14.04) so far, and have not checked the other two fancharts.

**Comment by Paul-Franklin** [comment]
https://github.com/gramps-project/gramps/commit/2f608f67e6a7e285402b32c9b311ca59539e2499

**Comment by enlite93** [comment]
Any update to this?

**Comment by enlite93** [comment]
What type of help do you want? 

As far as I can tell the three issue relating to the behavior of the tree when empty are not just for this chart/view and should just be reported.

For the Gtk issues with labels not displaying you'll need someone familiar with GUI design.

This would be good to have in Gramps 5.0.

**Comment by sam-m888** [comment]
> What type of help do you want? 

Gtk issues with the labels on the chart.

**Comment by prculley** [comment]
I just downloaded and looked at the chart and it looks great.  I see a few deprecation warnings, but no killer bugs on my Windows installation.  I'm not at all sure what you mean when you say "Gtk issues with the labels on the chart".  The only labels I see that are blank appear to be by design; I assume you mean the sectors in the graph where there isn't enough room to display a label.  (fanchart.py line 636).

BTW, some of the deprecation warnings were fixed in a recent PR I pushed [https://github.com/gramps-project/gramps/pull/268]

**Comment by sam-m888** [comment]
> I'm not at all sure what you mean when you say "Gtk issues with the labels on the chart". 

The two mentioned above in https://github.com/gramps-project/gramps/pull/222#issuecomment-246901153

> The only labels I see that are blank appear to be by design; I assume you mean the sectors in the graph where there isn't enough room to display a label. (fanchart.py line 636).

Yes I was taking about the blank **"sectors"** , which labels(that you are calling labels) are you writing about?

Will rebase and test this over the weekend after your PR #268 has been accepted.

**Comment by prculley** [comment]
To be as clear as possible;
The issues (Gramps.py:3058): Gtk-CRITICAL and (Gramps.py:3697): Pango-WARNING did not appear when I tested the chart.  So it is either Gtk version or platform dependent, or something got patched since those issues were posted.

The blank sectors of the chart that seems like they should contain names, if there was room for them, are the only "missing labels" I could find.  When I tracked them down, I found that the the 2-way fanchart inherits much of the code from fanchart.py.  That code includes a `wrap_truncate_layout()` method which trims the names down to what will fit in the sector.  But that code gives up if the sector is too small and doesn't include the label at all.  The line in question is fanchart.py line 636.

IMHO, this PR is ready to include as it is.

BTW, I did find a small bug caused in part by the fan chart code.  Seems if you run the three types of fan chart (normal, descendant, 2-way) you end up with three copies of new name formats (Edit/Preferences/Display Name format:  then Edit the name formats) and see 'Surname' and ''Given Suffix" more than once.  I think the bug has been there forever, I have a small patch for gen/display/name.py that corrects it I can push if you think it is worth fixing.

**Comment by Paul-Franklin** [comment]
" I have a small patch for gen/display/name.py that corrects
it. I can push if you think it is worth fixing."

In general I am in favor of any bug being fixed, if there is a
patch for it, especially ones in areas of gramps which might
not be heavily exercised (like custom name formats or the
name editor).

It is somewhat of a personal preference in style though,
so it's really up to you whatever you decide to do.

I always wonder when I am touching something which
gets heavily used (like the name displayer) whether I
am introducing something untoward in the process of
trying to fix whatever I am trying to fix, but perhaps that
is just me.

**Comment by sam-m888** [comment]
@prculley Rebased on latest master  and yes one of the warnings for me has gone only:
`(Gramps.py:4080): Pango-WARNING **: Invalid UTF-8 string passed to pango_layout_set_text()`
remains, this is on Ubuntu 14.04 so it maybe an issue with an old pango?
```
(geps/gep-030-FanChart2Way)]$ ./Gramps.py -v
Gramps Settings:
----------------
 python    : 3.4.3
 gramps    : 5.0.0-alpha1-1aa66ab

 gtk++     : 3.10.8
 pygobject : 3.12.0
 pango     : 1.36.3
 cairo     : 1.13.1
 pycairo   : 1.10.0
 osmgpsmap : 1.0
 GExiv2    : 0.10
 ICU       : 52.1
 PyICU     : 1.9.2
 o.s.      : linux
 kernel    : 3.13.0-36-generic

```

> 
The blank sectors of the chart that seems like they should contain names, if there was room for them, are the only "missing labels" I could find. When I tracked them down, I found that the the 2-way fanchart inherits much of the code from fanchart.py. That code includes a wrap_truncate_layout() method which trims the names down to what will fit in the sector. But that code gives up if the sector is too small and doesn't include the label at all. The line in question is fanchart.py line 636.

Thanks for the explanation very clear.

**Comment by sam-m888** [comment]
> IMHO, this PR is ready to include as it is.

That would be really great :+1:

**Comment by uxcq** [comment]
If I had the rights to merge this I would. Looks like you might be waiting a while!

**Comment by prculley** [comment]
I've attempted to look into the causes of the Pango error.  It seems likely that the problem comes when we try to truncate a name to fit the box.  The code appears to be correct on our side, and I could not get it to generate an error no matter what utf-8 strings with non-ASCII characters I put where they were truncated.  So I suspect that it is indeed a Pango version issue.  My Pango version is 1.40.1.

Unfortunately, I could find no obvious bug reports or commits in the Pango code that might indicate a fix.

Testing for this will be difficult, it depends on the location of a non-ASCII character versus the truncation point, which in turn depends on installed fonts, window size, # of people in the display etc.

Sam, I don't know if you have access to any other machines of VMs for running tests with your db, it would be interesting to see if you could replicate this bug elsewhere.  Or if you would be willing to share your db with me, I could try to replicate it.

**Comment by uxcq** [comment]
@prculley Looks like the individual with the error shown is "**I0914**" _Dennis Сергеев_ from the standard Family Tree "[example.gramps](https://gramps-project.org/wiki/index.php?title=Example.gramps)" that comes with Gramps.

**Comment by sam-m888** [comment]
Hi Paul,

Yes for testing I used the standard example.gramps[1] that comes with
Gramps, as shown in the screenshots above, I confirm the individual with
the error shown is "I0914" Dennis Сергеев.

Me testing in another VM will have to wait another week due to real life :(

Thanks for testing PR222: Gep-030 FanChart2Way.

Cheers
Sam

[1] https://gramps-project.org/wiki/index.php?title=Example.gramps

**Comment by prculley** [comment]
I'm not exactly loaded with extra time at the moment, but I tried for a while to get this to fail with the example.gramps db.  No luck.  Since I can only see a small portion of the chart in the screen capture above, I was unable to determine exactly the configuration to get it to display the same.

I think I need the center person, chart config for Max descendant generations, and Max ancestor generations, Fan chart distribution and any other details that might effect the chart display (filters?).

If I can get the exact layout you had with no error, I think we can chalk it up to a Gtk bug.

**Comment by eno93** [comment]
> I think I need the center person, 

Believe it's [I0044  Garner von Zieliński Lewis Anderson Sr ](https://cloud.githubusercontent.com/assets/8924713/18499751/2b7820f4-7a85-11e6-9a6f-b362001ac490.png)

Who happens to be the default person for example.gramps 

Then it looks like @sam-m888 clicked on I0174 Josephine Pelletier

**Comment by Nick-Hall** [comment]
I can reproduce the pango warnings.  I'll investigate further.

**Comment by Nick-Hall** [comment]
The problem is that the `set_text` method takes the length of the utf-8, not the length of the unicode as the second parameter.  So you need:
```
txt = 'Сергеев'
layout.set_text(txt, len(txt.encode('utf-8')))
```

**Comment by QuLogic** [comment]
Can you just pass -1?

File: gramps/gui/widgets/fanchart.py [hunk 1]
```diff
@@ -534,166 +620,138 @@ def wrap_truncate_layout(self, layout, font, width_pixels):
...
             layout.set_text(layout.get_text(), layout.get_line(0).length)
 
+        #2. we check if height is ok
+        w, h = layout.get_pixel_size()
+        if h > height_pixels:
+            if tryrescale:
+                #try to reduce the height
+                fontsize = max(height_pixels / h * font.get_size() /1.1, font.get_size()/2.0)
+                font.set_size(fontsize)
+                layout.set_text(all_text_backup, len(all_text_backup)) # reducing the height allows for more characters
```
Comment (eno93) [inline gramps/gui/widgets/fanchart.py]: Change this to:
`layout.set_text(all_text_backup, len(all_text_backup.encode('utf-8'))) # reducing the height allows for more characters`

as mentioned by @Nick-Hall  in https://github.com/gramps-project/gramps/pull/222#issuecomment-262861239

**Comment by sam-m888** [comment]
@Nick-Hall @eno93 That fixes the issue for me, Thank you. 

2016-11-25 16_30_37-ubuntu-16 04 1-grampsdev-amd64 running - oracle vm virtualbox

**Comment by Nick-Hall** [comment]
@QuLogic Yes, -1 also works.

**Comment by uxcq** [comment]
Excellent, now it just needs to be merged.  Anyone with rights to do so?

**Comment by Nick-Hall** [comment]
Merged fec5d532d1a691bec53c12f6ffff0eb7c4a3d40d.

**Comment by sam-m888** [comment]
Thank you @bastienjacquet  for creating this.

This will be part of Gramps 5.0 when released.

**Comment by QuLogic** [comment]
@Nick-Hall Why do you not use the Merge button on GitHub?

**Comment by Nick-Hall** [comment]
@QuLogic I always test the PR in a local branch.  It is then very easy to merge and push from the command line.  If the commit hash is unchanged, then GitHub will automatically mark the PR as merged.  However in this case I did a rebase before the merge.  The commit hash changed, so I had to close the PR manually.

For small changes, a rebase followed by a ff merge creates a neater history.

**Comment by QuLogic** [comment]
If the "Allow edits from maintainers" option on the right is ticked (I don't know if it is here, but it is by default), then you could push the rebase to the PR and replace it just before merging.

Of course, that's not a good idea if you don't plan on merging right away and don't notify the original PR opener, who might be a bit confused by it, but it does keep the PR status "semantically correct".

**Comment by Nick-Hall** [comment]
How do I know if I have permission to push to someone else's fork?  I can't see any "Allow edits from maintainers" option.

**Comment by QuLogic** [comment]
You can't see the check mark, but at the bottom just above the status there should be a message: "Add more commits by pushing to the `xyz` branch on `owner/gramps`." You'd need to add that person's repo as a remote and push to their `xyz` branch if it's available.

**Comment by Nick-Hall** [comment]
Thanks. I can see that on three out of the four open pull requests.

**Comment by sam-m888** [comment]
I can see "Allow edits from maintainers" is not ticked, as github planned it that way for PR's opened before that feature was added to github:

  * 2016-09-08 : [Improving collaboration with forks](https://github.com/blog/2247-improving-collaboration-with-forks) 

> Pull requests created before the option was available will default to not allowing collaboration, but contributors

**Comment by Nick-Hall** [comment]
Thanks for the link.

**Comment by eno93** [comment]
Here's a quick preview for those you not able to run the latest and greatest.

The new **FanChart2Way** 
fanchart2way-new

Due to lots of modifications to the underlying shared fan charts code the old charts look alot better.

The standard **Fanchart** original on the left and updated version on the right, cleaner lines.
fanchart-original and updated

The original **Descendant Fan** chart looks a bit drab, text is outside of and too large for each person box.

descendantfan-original

Updated **Descendant Fan** chart. Notice that each person name text  is clearer and appears on two lines, as well as appearing in the correct orientation depending on what side the person is.
descendantfan-updated

What's missing, no zoom button and printing the chart is a bit wonky 👍 

A great addition to Gramps.

**Comment by Nick-Hall** [comment]
We could consider backporting this to the gramps42 branch.

**Comment by uxcq** [comment]
@Nick-Hall Personally I'd say no to back-porting this as giving up-graders some visible reason to go to Gramps 5 which is a good thing taking into account all the changes that have happened.  

Does Gramps use the blog for more than release announcements?

**Comment by Paul-Franklin** [comment]
I agree.

I too have lots of things which I'd like users to be using.
I tried to get that to happen by agitating for a 4.3.0 but
it was decided we should put our work into getting 5.0.0
ready to go.  Fine.  Let's keep working on 5.0.0 and not
allow ourselves to get distracted on non-5.0.0 things.

**Comment by CedricGhestem** [comment]
@sam-m888 : 
your FanChart2Way is (my) best useful and used tree so far ; 
how difficult would it be (for you) to add these features : 
Allow export in vector file (svg or pdf / allow to keep text edition) 
Allow the text orientation (along the circle radius or perpendicular to it)
Allow to add more informations like dates and places of birth / death
Allow the text/font size : as far as we could export to vector format, who cares about the resolution...), to avoid blank cells.
Thanks a lot for what you've done !
Best regards

**Comment by giotodibondone** [comment]
@CedricGhestem Feature request can please be requested over on the Gramps [bugtracker](https://gramps-project.org/bugs/my_view_page.php) instead of on a 9 year old closed pull request, this is the wrong place!

Yes one of my favorite views also.

**Comment by CedricGhestem** [comment]
@giotodibondone : done, thanks !

**Comment by bastienjacquet** [comment]
At least I got a notification that the view I did a loooong time ago is useful to many.
Thanks for that to both of you !

**Comment by giotodibondone** [comment]
>  Thanks for that to both of you !

😊 this is one of my most used views and was already part of Gramps when I found it. thanks 🥇 

Feature request [13599](https://gramps-project.org/bugs/view.php?id=13599) submitted @CedricGhestem

**Comment by CedricGhestem** [comment]
@bastienjacquet   : congrats for your new (absorbing I guess) baby and job !
Do you think the features I suggested would be accessible ?
Or by you or by an other programmer you know I could ask ?
The 3 friends I asked in my (real life) relatives don't know Gramps nor Python enough.
(as a reminder : 
Allow export in vector file (svg or pdf / allow to keep text edition)
Allow the text orientation (along the circle radius or perpendicular to it)
Allow to add more informations like dates and places of birth / death
Allow the text/font size : as far as we could export to vector format, who cares about the resolution...), to avoid blank cells.)
Thanks a lot anyway !

**Comment by bastienjacquet** [comment]
> @bastienjacquet : congrats for your new (absorbing I guess) baby and job !

Thanks. It was 10 years ago, got 3 and started a company since then. ;)
> Do you think the features I suggested would be accessible ? 

>  Allow export in vector file (svg or pdf / allow to keep text edition) 

 never did that, but seems doable: https://stackoverflow.com/questions/69785424/how-to-save-svg-in-pycairo-w-o-context-manager
 I haven't checked if that would create related trouble (is svg cairo canvas available through gtk on mac/windows/linux ?) 
> Allow the text orientation (along the circle radius or perpendicular to it)

Easy. I could do it in the next months 
> Allow to add more informations like dates and places of birth / death 

I remember I did that on my personnal local? fork. But it raise many questions like where, how lengthy and what to show for unexact dates  
> Allow the text/font size : as far as we could export to vector format, who cares about the resolution...), to avoid blank cells.) 

Allow for font choice is easy. Allowing for base size is easy. Allowing for custom minimal size before cut-of is easy. 
But this does not make sense and cannot work without switching to svg first
> Thanks a lot anyway !

You're welcome. I was really just sharing something I made for myself. :)

**Comment by emyoulation** [comment]
If you have not seen it on the Gramp Discourse-powered community support form, a new experimental FamilyTreeView (FTV) is adding a SVG export and a PDF print for the view. 

Maybe some of that work could save some duplication of effort.

https://gramps.discourse.group/t/new-familytreeview-addon/6641

You have Chart View and Gramplet variants of the 2-way Fan. Since you already have SVG/PDF output, perhaps you could adapt to a Report version too? Maybe with a way to re‐use let the Report, gramplet and view select the configuration from the others? (Having the Gramplet and Chart View have independent  configurations is nice. We can use the view and gramplets to contrast different data using different background color schemes: age vs. gender.) After interactively configuring generations, colors, generations in the Charts view, it would be nice to re-use those setting for 2Way Fan chart reports for a series of People/Families.

Currently, there is only an Ancestor Fan in Reports.

Also, a 2-way Fan that centers on ancestors and descendants of a "Family" (the father and mother) instead of a "Person" would allow generating fewer charts per book of reports.

**Pull request closed without merging**